File: Day #763/solution.go
// Day 763: Sliding window maximum via a monotonic decreasing deque of indices.
// Time: O(n), Space: O(k). Prints each window max as it is computed.
package main

import "fmt"

func slidingMax(a []int, k int) {
	dq := []int{} // indices, values decreasing
	out := []int{}
	for i, x := range a {
		for len(dq) > 0 && a[dq[len(dq)-1]] <= x {
			dq = dq[:len(dq)-1]
		}
		dq = append(dq, i)
		if dq[0] <= i-k {
			dq = dq[1:]
		}
		if i >= k-1 {
			out = append(out, a[dq[0]])
		}
	}
	fmt.Println(out)
}

func main() {
	slidingMax([]int{10, 5, 2, 7, 8, 7}, 3) // [10 7 8 8]
}

File: Day #763/solution.cpp
// Day 763: Sliding window maximum via a monotonic decreasing deque of indices.
// Time: O(n), Space: O(k). Prints each window max as it is computed.
#include <bits/stdc++.h>
using namespace std;

void slidingMax(vector<int>& a, int k) {
    deque<int> dq;                 // indices, values decreasing
    vector<int> out;
    for (int i = 0; i < (int)a.size(); ++i) {
        while (!dq.empty() && a[dq.back()] <= a[i]) dq.pop_back();
        dq.push_back(i);
        if (dq.front() <= i - k) dq.pop_front();   // drop out-of-window
        if (i >= k - 1) out.push_back(a[dq.front()]);
    }
    cout << "[";
    for (size_t i = 0; i < out.size(); ++i)
        cout << out[i] << (i + 1 < out.size() ? ", " : "");
    cout << "]\n";
}

int main() {
    vector<int> a = {10, 5, 2, 7, 8, 7};
    slidingMax(a, 3);   // [10, 7, 8, 8]
    return 0;
}
